### PR TITLE
Accept any Mapping as output of a plugin widget

### LINKED ISSUE #7249: Plugins that return a layer data tuple with a `DeprecatingDict` fail
### 🐛 Bug Report

When a plugin returns a layer data tuple where the second element is a `DeprecatingDict` instead of a `dict`, the layer is not validated correctly and raises the following error message.

```
TypeError: Not a valid list of layer data tuples!
```

This exception is raised from `napari.utils.misc.ensure_layer_data_tuple`, which explicitly checks for a `dict`.

This error is most likely to occur when a plugin implementation calls `Layer.as_layer_data_tuple` to initialize the mapping of attributes, then modify that mapping, and return the modified mapping as part of a new tuple.

### 💡 Steps to Reproduce

Here is a test that can be added to `test_magicgui` to reproduce the error.

```python
def test_magicgui_from_layer_data_tuple(make_napari_viewer):
    viewer = make_napari_viewer()

    @magicgui
    def from_layer_data_tuple(image: Image) -> types.LayerDataTuple:
        return image.as_layer_data_tuple()

    viewer.window.add_dock_widget(from_layer_data_tuple)
    image = Image(data=np.zeros((10, 10)))

    from_layer_data_tuple(image)

    assert len(viewer.layers) == 1
    assert isinstance(viewer.layers[0], Image)
    assert viewer.layers[0].source.widget == from_layer_data_tuple
```

### 💡 Expected Behavior

I expect this to run without error.

### 🌎 Environment

napari: 0.4.17rc4.dev1063+ge1be2a723.d20240904
Platform: macOS-14.6.1-arm64-arm-64bit
System: MacOS 14.6.1
Python: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:51:49) [Clang 16.0.6 ]
Qt: 5.15.11
PyQt5: 5.15.10
NumPy: 2.0.0
SciPy: 1.14.0
Dask: 2024.6.2
VisPy: 0.14.3
magicgui: 0.8.3
superqt: 0.6.7
in-n-out: 0.2.1
app-model: 0.2.8
npe2: 0.7.6

OpenGL:
  - GL version:  2.1 Metal - 88.1
  - MAX_TEXTURE_SIZE: 16384
  - GL_MAX_3D_TEXTURE_SIZE: 2048

Screens:
  - screen 1: resolution 1512x982, scale 2.0

Optional:
  - numba: 0.60.0
  - triangle not installed
  - napari-plugin-manager: 0.1.0a2

Settings path:
  - /Users/asweet/Library/Application Support/napari/napari-dev_bed8b9b3d61b11a9560a577ec715d63283498f35/settings.yaml
Plugins:
  - napari: 0.4.17rc4.dev1063+ge1be2a723.d20240904 (81 contributions)
  - napari-console: 0.0.9 (0 contributions)
  - napari-svg: 0.2.0 (2 contributions)

### 💡 Additional Context

See the [original post on Zulip](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/DeprecatingDict.20not.20accepted.20in.20LayerDataTuple)

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues

closes #7249

# Description

### CODE REVIEW

File: napari/_tests/test_magicgui.py [hunk 1]
```diff
@@ -346,3 +346,21 @@ def test_layers_populate_immediately(make_napari_viewer):
...
+
+
+def test_from_layer_data_tuple(make_napari_viewer):
+    """Test that a function returning a layer data tuple runs without error."""
+    viewer = make_napari_viewer()
+
+    @magicgui
+    def from_layer_data_tuple() -> types.LayerDataTuple:
+        data = np.random.rand(10, 10)
+        meta = {'name': 'test_image'}
```
Comment (andy-sweet) [inline napari/_tests/test_magicgui.py]: Given that you are creating a `dict` here, I don't think this test fails on main.

I'd assume that the return values should be initialized with `Layer.as_layer_data_tuple` to check internal consistency within napari like in the reproducer in the related issue. That's how the bug arose in practice and it seems like a natural usage.

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Approving the implementation, but I think the test needs to change to reproduce the issue this should fix.

File: napari/_tests/test_magicgui.py [hunk 2]
```diff
@@ -346,3 +347,21 @@ def test_layers_populate_immediately(make_napari_viewer):
...
     assert len(labels_layer.choices) == 1
+
+
+def test_from_layer_data_tuple(make_napari_viewer):
+    """Test that a function returning a layer data tuple runs without error."""
+    viewer = make_napari_viewer()
+
+    @magicgui
+    def from_layer_data_tuple() -> types.LayerDataTuple:
+        data = np.random.rand(10, 10)
```
Comment (andy-sweet) [inline napari/_tests/test_magicgui.py]: Nit: why use `random` here instead of `zeros`?

File: napari/_tests/test_magicgui.py [hunk 3]
```diff
@@ -346,3 +347,21 @@ def test_layers_populate_immediately(make_napari_viewer):
     assert not len(labels_layer.choices)
     viewer.window.add_dock_widget(labels_layer)
     assert len(labels_layer.choices) == 1
+
+
+def test_from_layer_data_tuple(make_napari_viewer):
```
Comment (andy-sweet) [inline napari/_tests/test_magicgui.py]: Nit: consider renaming or modifying the comment to indicate this is specifically about testing a `_DeprecatingDict`.

**Review by andy-sweet — APPROVED** [review]
Approving because this fixes the issue reported through natural plugin usage.

I'm unsure if `Mapping` is enough for internal usage partly because we don't have type checks through the code base. But I think it's OK to try it, then expand it as needed.

Given that napari's and npe2's `types` modules both indicate they expect the second element of a layer data tuple to be a `dict`, I am considering re-implementing `_DeprecatingDict` to subclass `dict` instead of `UserDict`. But that is a separate PR that is not inconsistent with this one.

Also, I'd prefer a test with `as_layer_data_tuple` because that is the public API that others may use, but explicitly testing the implementation detail of `_DeprecatingDict` is good enough for now.

**Comment by Czaki** [comment]
> Given that napari's and npe2's `types` modules both indicate they expect the second element of a layer data tuple to be a `dict`, I am considering re-implementing `_DeprecatingDict` to subclass `dict` instead of `UserDict`. But that is a separate PR that is not inconsistent with this one.

I strongly prefer to update types in both places to `Mapping`. Subclassing dict is not a good idea.

**Pull request merged**